### PR TITLE
Add TTL 74166

### PR DESCRIPTION
**Pull request by btb**
split off from #1461

### CODE REVIEW

**Review by maehne — DISMISSED** [review]
@btb: Thanks, looks good to me!

File: src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+import com.cburch.logisim.instance.InstancePainter;
+import com.cburch.logisim.instance.InstancePoker;
+import com.cburch.logisim.instance.InstanceState;
+import com.cburch.logisim.instance.StdAttr;
+import com.cburch.logisim.util.GraphicsUtil;
+import java.awt.Color;
+import java.awt.Graphics2D;
+import java.awt.event.MouseEvent;
+
+public class Ttl74166 extends AbstractTtlGate {
```
Comment (maehne) [inline src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java:24]: Please, add a docstring describing the purpose of the class.

File: src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    public void mousePressed(InstanceState state, MouseEvent e) {
+      isPressed = isInside(state, e);
+    }
+
+    @Override
+    public void mouseReleased(InstanceState state, MouseEvent e) {
+      if (!state.getAttributeValue(TtlLibrary.DRAW_INTERNAL_STRUCTURE)) return;
+      if (isPressed && isInside(state, e)) {
+        final var index = getIndex(state, e);
+        System.out.println(index);
```
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java:98]: Please remove.

File: src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        final var myState = (ShiftRegisterData) state.getData();
+        if (myState == null) return;
+        if (myState.get(index).isFullyDefined())
```
Comment (MarcinOrlowski) [inline src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java:101]: Please add `{}` to code blocks

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

File: src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+import com.cburch.logisim.instance.InstancePoker;
+import com.cburch.logisim.instance.InstanceState;
+import com.cburch.logisim.instance.StdAttr;
+import com.cburch.logisim.util.GraphicsUtil;
+import java.awt.Color;
+import java.awt.Graphics2D;
+import java.awt.event.MouseEvent;
+
+/**
+ * TTL 74x166 8-bit parallel-to-serial shift register with clear
```
Comment (maehne) [inline src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java:25]: Please, be a bit more specific: "clear" -> "async clear"

File: src/main/resources/resources/logisim/strings/std/std.properties [hunk 5]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: hex inverter (schmitt-trigger)
 TTL74161 = 74161: 4-bit sync counter with async clear
 TTL74163 = 74163: 4-bit sync counter with sync clear
 TTL74165 = 74165: 8-bit parallel-to-serial shift register
+TTL74166 = 74166: 8-bit parallel-to-serial shift register with clear
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std.properties:777]: "clear" -> "async clear"

File: src/main/resources/resources/logisim/strings/std/std_nl.properties [hunk 6]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: 6-voudige inverter (schmitt-trigger)
 TTL74161 = 74161: synchrone 4-bits binaire teller met asynchrone clear
 TTL74163 = 74163: synchrone 4-bits binaire teller met synchrone clear
 TTL74165 = 74165: 8-bits parallel-naar-serieel schuifregister
+TTL74166 = 74166: 8-bits parallel-naar-serieel schuifregister met clear
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_nl.properties:777]: "clear" -> "asynchrone clear"

File: src/main/resources/resources/logisim/strings/std/std_pl.properties [hunk 7]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: falownik sześciokątny (schmitt-trigger)
 # ==> TTL74161 =
 # ==> TTL74163 =
 TTL74165 = 74165: 8-bitowy rejestr przesunięcia równoległego do szeregowego
+TTL74166 = 74166: 8-bitowy rejestr przesunięcia równoległego do szeregowego z wyczyść
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_pl.properties:777]: If possible, please be more precise to state that the clear is asynchronous.

@MarcinOrlowski: Could you maybe help?

File: src/main/resources/resources/logisim/strings/std/std_ru.properties [hunk 8]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: шестнадцатеричный инвертор (шмитт-
 TTL74161 = 74161: 4-bit sync counter with clear
 # ==> TTL74163 =
 TTL74165 = 74165: 8-битный регистр параллельного последовательного сдвига
+TTL74166 = 74166: 8-битный регистр параллельного последовательного сдвига с очистить
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_ru.properties:777]: If possible, please be more precise to state that the clear is asynchronous.

File: src/main/resources/resources/logisim/strings/std/std_pt.properties [hunk 9]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: inversor hexadecimal (schmitt-trigger)
 TTL74161 = 74161: 4-bit sync counter with clear
 # ==> TTL74163 =
 TTL74165 = 74165: Registo de deslocação paralela para série de 8-bit
+TTL74166 = 74166: Registo de deslocação paralela para série de 8-bit com claro
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_pt.properties:777]: If possible, please be more precise to state that the clear is asynchronous.

File: src/main/resources/resources/logisim/strings/std/std_de.properties [hunk 10]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: Sechsfach Invertierer (Schmitt-Trigger)
 TTL74161 = 74161: synchroner 4-bit-Zähler mit Clear
 TTL74163 = 74163: synchroner 4-bit-Zähler mit synchronem Clear
 TTL74165 = 74165: 8-Bit Parallel zu Serielles Schieberegister
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_de.properties:776]: "Parallel zu Serielles" -> "parallel-zu-seriell"

File: src/main/resources/resources/logisim/strings/std/std_de.properties [hunk 11]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: Sechsfach Invertierer (Schmitt-Trigger)
...
+TTL74166 = 74166: 8-Bit Parallel zu Serielles Schieberegister mit Clear
```
Comment (maehne) [inline src/main/resources/resources/logisim/strings/std/std_de.properties:777]: "Parallel zu Serielles" -> "parallel-zu-seriell"
"Clear" -> "asynchronem Clear"

**Review by maehne — CHANGES_REQUESTED** [review]
Thanks for addressing my comments! Looks mostly good to me! Though, it would be nice if you could address @MarcinOrlowski's and mine additional suggestions.

File: src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java (hunk 2 above)
Comment (btb) [reply src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java]: done

File: src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java (hunk 3 above)
Comment (btb) [reply src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java]: done

File: src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java (hunk 4 above)
Comment (btb) [reply src/main/java/com/cburch/logisim/std/ttl/Ttl74166.java]: done

File: src/main/resources/resources/logisim/strings/std/std_nl.properties (hunk 6 above)
Comment (btb) [reply src/main/resources/resources/logisim/strings/std/std_nl.properties]: done

File: src/main/resources/resources/logisim/strings/std/std_pl.properties (hunk 7 above)
Comment (btb) [reply src/main/resources/resources/logisim/strings/std/std_pl.properties]: done

File: src/main/resources/resources/logisim/strings/std/std_ru.properties (hunk 8 above)
Comment (btb) [reply src/main/resources/resources/logisim/strings/std/std_ru.properties]: done

File: src/main/resources/resources/logisim/strings/std/std_de.properties (hunk 10 above)
Comment (btb) [reply src/main/resources/resources/logisim/strings/std/std_de.properties:776]: will fix in upcoming PR to fix #1472

File: src/main/resources/resources/logisim/strings/std/std_de.properties (hunk 11 above)
Comment (btb) [reply src/main/resources/resources/logisim/strings/std/std_de.properties]: done

File: src/main/resources/resources/logisim/strings/std/std_es.properties [hunk 12]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: inversor hexagonal (schmitt-trigger)
 TTL74161 = 74161: 4-bit sync counter with clear
 # ==> TTL74163 =
 TTL74165 = 74165: Registro de desplazamiento paralelo a serie de 8 bits
+# ==> TTL74166 = 74166: Registro de desplazamiento paralelo a serie de 8 bits - FIXME: "with asynchronous clear"
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_es.properties:777]: Please remove anything but the key. So it should be just

```ini
# ==> TTL74166
```

applies to all the cases below as well. One can always use [trans-tool](https://github.com/MarcinOrlowski/trans-tool) to get reference strings during works on their translation

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

File: src/main/resources/resources/logisim/strings/std/std_es.properties (hunk 12 above)
Comment (btb) [reply src/main/resources/resources/logisim/strings/std/std_es.properties:777]: existing empty translations all include the equal sign as well, i.e.
```
# ==> TTL74166 =
```

should it not be done that way?

Comment (MarcinOrlowski) [reply src/main/resources/resources/logisim/strings/std/std_es.properties:777]: It's not the case for all the translation files, hence my request.

File: src/main/resources/resources/logisim/strings/std/std_pl.properties [hunk 13]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: falownik sześciokątny (schmitt-trigger)
 # ==> TTL74161 =
 # ==> TTL74163 =
 TTL74165 = 74165: 8-bitowy rejestr przesunięcia równoległego do szeregowego
+# ==> TTL74166 = 74166: 8-bitowy rejestr przesunięcia równoległego do szeregowego - FIXME: "with asynchronous clear"
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_pl.properties:777]:
```ini
TTL74166 = 74166: 8-bitowy rejestr przesunięcia równoległego do szeregowego z kasowaniem asynchronicznym
```

File: src/main/resources/resources/logisim/strings/std/std_pt.properties [hunk 14]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: inversor hexadecimal (schmitt-trigger)
 TTL74161 = 74161: 4-bit sync counter with clear
 # ==> TTL74163 =
 TTL74165 = 74165: Registo de deslocação paralela para série de 8-bit
+# ==> TTL74166 = 74166: Registo de deslocação paralela para série de 8-bit - FIXME: "with asynchronous clear"
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_pt.properties:777]: Please leave just a key.

```ini
# ==> TTL74166 =
```

File: src/main/resources/resources/logisim/strings/std/std_ru.properties [hunk 15]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: шестнадцатеричный инвертор (шмитт-
 TTL74161 = 74161: 4-bit sync counter with clear
 # ==> TTL74163 =
 TTL74165 = 74165: 8-битный регистр параллельного последовательного сдвига
+# ==> TTL74166 = 74166: 8-битный регистр параллельного последовательного сдвига - FIXME: "with asynchronous clear"
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_ru.properties:777]: Please leave just a key.

```ini
# ==> TTL74166 =
```

File: src/main/resources/resources/logisim/strings/std/std_nl.properties [hunk 16]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: 6-voudige inverter (schmitt-trigger)
 TTL74161 = 74161: synchrone 4-bits binaire teller met asynchrone clear
 TTL74163 = 74163: synchrone 4-bits binaire teller met synchrone clear
 TTL74165 = 74165: 8-bits parallel-naar-serieel schuifregister
+# ==> TTL74166 = 74166: 8-bits parallel-naar-serieel schuifregister - FIXME: "with asynchronous clear"
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_nl.properties:777]: Please leave just a key.

```ini
# ==> TTL74166 =
```

File: src/main/resources/resources/logisim/strings/std/std_it.properties [hunk 17]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414: inverter esagonale (schmitt-trigger)
 # ==> TTL74161 =
 # ==> TTL74163 =
 TTL74165 = 74165: registro di spostamento da parallelo a seriale a 8 bit
+# ==> TTL74166 = 74166: registro di spostamento da parallelo a seriale a 8 bit - FIXME: "with asynchronous clear"
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_it.properties:777]: Please leave just a key.

```ini
# ==> TTL74166 =
```

File: src/main/resources/resources/logisim/strings/std/std_fr.properties [hunk 18]
```diff
@@ -774,6 +774,7 @@ TTL7414 = 7414 : onduleur hexagonal (schmitt-trigger)
 TTL74161 = 74161: 4-bit sync counter with clear
 # ==> TTL74163 =
 TTL74165 = 74165 : registre à décalage parallèle-à-série 8 bits
+# ==> TTL74166 = 74166 : registre à décalage parallèle-à-série 8 bits - FIXME: "with asynchronous clear"
```
Comment (MarcinOrlowski) [inline src/main/resources/resources/logisim/strings/std/std_fr.properties:777]: Please leave just a key.

```ini
# ==> TTL74166 =
```

**Review by MarcinOrlowski — CHANGES_REQUESTED** [review]

**Comment by maehne** [comment]
@MarcinOrlowski: What still needs to be done before we can merge this PR?

**Comment by BFH-ktt1** [comment]
@MarcinOrlowski : What is still blocking this PR?

**Review by MarcinOrlowski — APPROVED** [review]

**Comment by MarcinOrlowski** [comment]
Sorry for the lag. Code looks OK for me.

**Pull request merged**